### PR TITLE
Fixing flipped approach

### PR DESCRIPTION
**Pull request by mlautman**
This fixes the flipped approach vector bug and removes a double negative that shouldn't have been there according to a reasonable interpretation of the variable names. 

I also added a rosparam for ideal_grasp_orientation_rpy

### CODE REVIEW

File: config/grasp_debug_level.yaml [hunk 1]
```diff
@@ -18,6 +18,7 @@ moveit_grasps:
     translation_x_score_weight: 0.0
     translation_y_score_weight: 0.0
     translation_z_score_weight: 4.0
+    grasp_approach_vector_in_parent_frame: [0, 0, 1]
```
Comment (davetcoleman) [inline config/grasp_debug_level.yaml]: this does not go in grasp_debug_level, please move this now along with lines 13-20

File: include/moveit_grasps/grasp_generator.h [hunk 2]
```diff
@@ -321,6 +321,9 @@ class GraspGenerator
   // Transform from frame of box to global frame
   Eigen::Affine3d object_global_transform_;
 
+  // Approach vector in parent frame
```
Comment (davetcoleman) [inline include/moveit_grasps/grasp_generator.h]: avoid having the exact comment as the variable name

File: include/moveit_grasps/grasp_generator.h [hunk 3]
```diff
@@ -321,6 +321,9 @@ class GraspGenerator
...
+  std::vector<double> grasp_approach_vector_in_parent_frame_;
```
Comment (davetcoleman) [inline include/moveit_grasps/grasp_generator.h]: great variable name

File: src/grasp_data.cpp [hunk 4]
```diff
@@ -111,18 +110,13 @@ bool GraspData::loadGraspData(const ros::NodeHandle& nh, const std::string& end_
...
-  rosparam_shortcuts::get(parent_name, child_nh, "grasp_pose_to_eef_transform", grasp_pose_to_eef_transform);
+  rosparam_shortcuts::get(parent_name, child_nh, "grasp_pose_to_eef_transform", grasp_pose_to_eef_pose_);
   rosparam_shortcuts::get(parent_name, child_nh, "grasp_padding_on_approach", grasp_padding_on_approach_);
   rosparam_shortcuts::get(parent_name, child_nh, "max_finger_width", max_finger_width_);
   rosparam_shortcuts::get(parent_name, child_nh, "min_finger_width", min_finger_width_);
 
   // Convert generic grasp pose to this end effector's frame of reference, approach direction for short
 
-  // Orientation
-  ROS_ASSERT(grasp_pose_to_eef_transform.size() == 6);
```
Comment (davetcoleman) [inline src/grasp_data.cpp]: isn't grasp_pose_to_eef_pose_ still size 6? enforce this

File: src/grasp_generator.cpp [hunk 5]
```diff
@@ -66,15 +66,16 @@ GraspGenerator::GraspGenerator(moveit_visual_tools::MoveItVisualToolsPtr visual_
...
   // Set ideal grasp pose (currently only uses orientation of pose)
   ideal_grasp_pose_ = Eigen::Affine3d::Identity();
-  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(M_PI / 2.0, Eigen::Vector3d::UnitZ()) *
-                      Eigen::AngleAxisd(-M_PI / 2.0, Eigen::Vector3d::UnitX());
+  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(-M_PI, Eigen::Vector3d::UnitX());
   ideal_grasp_pose_.translation() = Eigen::Vector3d(0, 0, 2.0);
 
   // TODO(davetcoleman): make this an option
-  visual_tools_->publishAxisLabeled(ideal_grasp_pose_, "IDEAL_GRASP_POSE");
+  visual_tools_->publishAxisLabeled(ideal_grasp_pose_, "ideal grasp orientation");
```
Comment (davetcoleman) [inline src/grasp_generator.cpp]: use underscores for floating rviz labels... its nicer

also pose is more compact than orientation, why relabel this?

File: src/grasp_generator.cpp [hunk 6]
```diff
@@ -66,15 +66,16 @@ GraspGenerator::GraspGenerator(moveit_visual_tools::MoveItVisualToolsPtr visual_
...
   rosparam_shortcuts::get(parent_name, nh_, "translation_y_score_weight", translation_y_score_weight_);
   rosparam_shortcuts::get(parent_name, nh_, "translation_z_score_weight", translation_z_score_weight_);
+  rosparam_shortcuts::get(parent_name, nh_, "grasp_approach_vector_in_parent_frame",
+                          grasp_approach_vector_in_parent_frame_);
 
   // Set ideal grasp pose (currently only uses orientation of pose)
   ideal_grasp_pose_ = Eigen::Affine3d::Identity();
-  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(M_PI / 2.0, Eigen::Vector3d::UnitZ()) *
-                      Eigen::AngleAxisd(-M_PI / 2.0, Eigen::Vector3d::UnitX());
+  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(-M_PI, Eigen::Vector3d::UnitX());
```
Comment (davetcoleman) [inline src/grasp_generator.cpp]: add lots of comments about how this was chosen and what its for

keep the old approach and make it a rosparam

something like...

int ideal_grasp_pose_type = 1 // Shelf picking
int ideal_grasp_pose_type = 2 // Bin picking

File: src/grasp_planner.cpp [hunk 7]
```diff
@@ -76,14 +76,22 @@ bool GraspPlanner::planAllApproachLiftRetreat(std::vector<GraspCandidatePtr>& gr
       return false;
 
     ROS_INFO_STREAM_NAMED("grasp_planner", "");
-    ROS_INFO_STREAM_NAMED("grasp_planner", "Attempting to plan cartesian grasp path #" << count++);
+    ROS_INFO_STREAM_NAMED("grasp_planner", "Attempting to plan cartesian grasp path #"
+                                               << count++ << ". " << grasp_candidates.size() << " remaining.");
```
Comment (davetcoleman) [inline src/grasp_planner.cpp]: nice

File: src/grasp_planner.cpp [hunk 8]
```diff
@@ -76,14 +76,22 @@ bool GraspPlanner::planAllApproachLiftRetreat(std::vector<GraspCandidatePtr>& gr
...
       ROS_INFO_STREAM_NAMED("grasp_planner", "Grasp candidate was unable to find valid cartesian waypoint path");
 
       grasp_it = grasp_candidates.erase(grasp_it);  // not valid
+
+      // If this was the last grasp candidate then we failed to find a valid one
+      if (grasp_candidates.size() == 0)
+      {
+        ROS_ERROR_STREAM_NAMED("grasp_planner", "No valid grasp plan possible");
+        return false;
+      }
```
Comment (davetcoleman) [inline src/grasp_planner.cpp]: i like the verbose_cartesian_filtering debug output, below. can you remove this return false to after that?

File: config/grasp_debug_level.yaml (hunk 1 above)
Comment (mlautman) [reply config/grasp_debug_level.yaml]: I think that renaming grasp_debug_level -> moveit_grasps_config addresses the issue of lines 13-20. I will talk to andy about the getting this transform out of the grasp_pose_to_eef_transform parameter

File: src/grasp_data.cpp (hunk 4 above)
Comment (mlautman) [reply src/grasp_data.cpp]: Already discussed offline

Comment (mlautman) [reply src/grasp_data.cpp]: Rosparam supports loading 6 values into an affine3d container and takes care of this check internally.

File: src/grasp_generator.cpp [hunk 9]
```diff
@@ -67,14 +67,24 @@ GraspGenerator::GraspGenerator(moveit_visual_tools::MoveItVisualToolsPtr visual_
...
+  rosparam_shortcuts::get(parent_name, nh_, "ideal_grasp_orientation_rpy", ideal_grasp_orientation_rpy);
+  ROS_ASSERT(ideal_grasp_orientation_rpy.size() == 3);
+
   // Set ideal grasp pose (currently only uses orientation of pose)
   ideal_grasp_pose_ = Eigen::Affine3d::Identity();
-  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(M_PI / 2.0, Eigen::Vector3d::UnitZ()) *
-                      Eigen::AngleAxisd(-M_PI / 2.0, Eigen::Vector3d::UnitX());
+  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(ideal_grasp_orientation_rpy[0], Eigen::Vector3d::UnitX());
+  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(ideal_grasp_orientation_rpy[1], Eigen::Vector3d::UnitY());
+  ideal_grasp_pose_ = ideal_grasp_pose_ * Eigen::AngleAxisd(ideal_grasp_orientation_rpy[2], Eigen::Vector3d::UnitZ());
```
Comment (mcevoyandy) [inline src/grasp_generator.cpp]: You could use *= operator here. Why change the order of rotations? Pretty sure this needs to stay `pose * Z * Y * X`
Also, is the 2.0 translation on the next line needed? Should it be a param?

Comment (mlautman) [reply src/grasp_generator.cpp]: The `ideal_grasp_orientation_rpy` variable is Roll Pitch Yaw which is XYZ

Comment (mlautman) [reply src/grasp_generator.cpp]: Verified

**Comment by mlautman** [comment]
already fixed

**Pull request closed without merging**